File: scripts/ingest-dramas.mjs
#!/usr/bin/env node
/**
 * One-time import: mydramalist_kdramas_v2.json -> Supabase public.dramas
 * Requires: SUPABASE_URL or NEXT_PUBLIC_SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY in env.
 * Loads .env and .env.local from project root if present.
 * Optional: embedding, umap_x, umap_y on each item (nullable in DB).
 * For embeddings/UMAP from .npy files, use: python scripts/ingest_dramas.py
 */

import { createClient } from "@supabase/supabase-js";
import { readFileSync, existsSync } from "fs";
import { fileURLToPath } from "url";
import { dirname, join } from "path";

const __dirname = dirname(fileURLToPath(import.meta.url));
const projectRoot = join(__dirname, "..");

function loadEnvFile(path) {
  if (!existsSync(path)) return;
  const content = readFileSync(path, "utf8");
  for (const line of content.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) continue;
    const eq = trimmed.indexOf("=");
    if (eq === -1) continue;
    const key = trimmed.slice(0, eq).trim();
    let value = trimmed.slice(eq + 1).trim();
    if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'")))
      value = value.slice(1, -1);
    if (!process.env[key]) process.env[key] = value;
  }
}

loadEnvFile(join(projectRoot, ".env"));
loadEnvFile(join(projectRoot, ".env.local"));

const SUPABASE_URL = process.env.SUPABASE_URL || process.env.NEXT_PUBLIC_SUPABASE_URL;
const SUPABASE_SERVICE_ROLE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY;

if (!SUPABASE_URL || !SUPABASE_SERVICE_ROLE_KEY) {
  console.error("Set SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY (or NEXT_PUBLIC_SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY)");
  process.exit(1);
}

const supabase = createClient(SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY);

function parseNum(val) {
  if (val == null) return null;
  const n = Number(val);
  return Number.isFinite(n) ? n : null;
}

function parseRating(val) {
  const n = parseNum(val);
  if (n == null) return null;
  return Math.min(10, Math.max(0, Math.round(n * 10) / 10));
}

function toVector(arr) {
  if (arr == null) return null;
  if (Array.isArray(arr) && arr.length > 0) return `[${arr.join(",")}]`;
  return null;
}

function mapRow(item) {
  const year = parseNum(item.year);
  const num_episodes = parseNum(item.episodes);
  const rating = parseRating(item.rating);
  return {
    title: item.title ?? "",
    original_title: item.original_title ?? null,
    media_type: item.media_type ?? "Korean Drama",
    year: year ?? null,
    num_episodes: num_episodes ?? null,
    rating: rating ?? null,
    description: item.description ?? null,
    image_url: item.image_url ?? null,
    link: item.link ?? null,
    genres: Array.isArray(item.genres) ? item.genres : [],
    tags: Array.isArray(item.tags) ? item.tags : [],
    main_actors: Array.isArray(item.main_actors) ? item.main_actors : [],
    embedding: toVector(item.embedding),
    umap_x: parseNum(item.umap_x) ?? null,
    umap_y: parseNum(item.umap_y) ?? null,
  };
}

const BATCH_SIZE = 100;

async function main() {
  const jsonPath = join(projectRoot, "mydramalist_kdramas_v2.json");
  console.log("Reading", jsonPath);
  const raw = readFileSync(jsonPath, "utf8");
  const items = JSON.parse(raw);
  if (!Array.isArray(items)) {
    console.error("JSON root must be an array");
    process.exit(1);
  }

  const rows = items.map(mapRow);
  console.log("Mapped", rows.length, "rows. Inserting in batches of", BATCH_SIZE);

  let inserted = 0;
  for (let i = 0; i < rows.length; i += BATCH_SIZE) {
    const batch = rows.slice(i, i + BATCH_SIZE);
    const { data, error } = await supabase.from("dramas").insert(batch).select("id");
    if (error) {
      console.error("Batch error at offset", i, error);
      throw error;
    }
    inserted += (data?.length ?? 0);
    console.log("Inserted", inserted, "/", rows.length);
  }

  console.log("Done. Total inserted:", inserted);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: scripts/ingest_dramas.py
#!/usr/bin/env python3
"""
Deprecated: use `npm run ingest` (scripts/ingest-all.mjs) for a single Node-based workflow.

This file is kept for reference or environments that cannot run Node; behavior matches
ingest-all.mjs (JSON + optional .npy files).
"""

import json
import os
import sys
from pathlib import Path

import numpy as np
import requests

PROJECT_ROOT = Path(__file__).resolve().parent.parent


def load_env():
    for name in (".env", ".env.local"):
        path = PROJECT_ROOT / name
        if not path.exists():
            continue
        with open(path) as f:
            for line in f:
                line = line.strip()
                if not line or line.startswith("#"):
                    continue
                if "=" in line:
                    key, _, value = line.partition("=")
                    key, value = key.strip(), value.strip().strip("'\"").strip()
                    if key and not os.environ.get(key):
                        os.environ[key] = value


def parse_num(val):
    if val is None:
        return None
    try:
        n = float(val)
        return n if np.isfinite(n) else None
    except (TypeError, ValueError):
        return None


def parse_rating(val):
    n = parse_num(val)
    if n is None:
        return None
    return min(10.0, max(0.0, round(n * 10) / 10))


def main():
    load_env()

    url = os.environ.get("SUPABASE_URL") or os.environ.get("NEXT_PUBLIC_SUPABASE_URL")
    key = os.environ.get("SUPABASE_SERVICE_ROLE_KEY")
    if not url or not key:
        print("Missing NEXT_PUBLIC_SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY in .env / .env.local", file=sys.stderr)
        sys.exit(1)

    url = url.rstrip("/")
    api = f"{url}/rest/v1"
    headers = {
        "apikey": key,
        "Authorization": f"Bearer {key}",
        "Content-Type": "application/json",
        "Prefer": "return=representation",
    }

    json_path = PROJECT_ROOT / "mydramalist_kdramas_v2.json"
    emb_path = PROJECT_ROOT / "drama_embeddings.npy"
    umap_path = PROJECT_ROOT / "drama_embeddings_2d.npy"

    # DB column must match; set EMBED_DIM in .env if your .npy uses a different dimension.
    EMBED_DIM = int(os.environ.get("EMBED_DIM", "384"))

    if not json_path.exists():
        print(f"Not found: {json_path}", file=sys.stderr)
        sys.exit(1)

    print("Loading", json_path)
    with open(json_path) as f:
        items = json.load(f)
    if not isinstance(items, list):
        print("JSON root must be an array", file=sys.stderr)
        sys.exit(1)

    embeddings = None
    if emb_path.exists():
        print("Loading", emb_path)
        embeddings = np.load(emb_path)
        if embeddings.ndim != 2:
            print(f"Expected 2D array, got shape {embeddings.shape}", file=sys.stderr)
            sys.exit(1)
        if embeddings.shape[1] != EMBED_DIM:
            print(f"Embedding dim is {embeddings.shape[1]}; DB expects {EMBED_DIM}. Alter dramas.embedding or set EMBED_DIM.", file=sys.stderr)
            sys.exit(1)
        if len(embeddings) != len(items):
            print(f"Length mismatch: embeddings {len(embeddings)} vs JSON {len(items)}", file=sys.stderr)
            sys.exit(1)
    else:
        print("No drama_embeddings.npy; embedding column will be null")

    umap_2d = None
    if umap_path.exists():
        print("Loading", umap_path)
        umap_2d = np.load(umap_path)
        if umap_2d.ndim != 2 or umap_2d.shape[1] != 2:
            print(f"Expected Nx2 array, got shape {umap_2d.shape}", file=sys.stderr)
            sys.exit(1)
        if len(umap_2d) != len(items):
            print(f"Length mismatch: umap {len(umap_2d)} vs JSON {len(items)}", file=sys.stderr)
            sys.exit(1)
    else:
        print("No drama_embeddings_2d.npy; umap_x/umap_y will be null")

    batch_size = 100
    inserted = 0

    for start in range(0, len(items), batch_size):
        batch_items = items[start : start + batch_size]
        rows = []
        for i, item in enumerate(batch_items):
            idx = start + i
            year = parse_num(item.get("year"))
            num_episodes = parse_num(item.get("episodes"))
            rating = parse_rating(item.get("rating"))

            row = {
                "title": item.get("title") or "",
                "original_title": item.get("original_title"),
                "media_type": item.get("media_type") or "Korean Drama",
                "year": int(year) if year is not None else None,
                "num_episodes": int(num_episodes) if num_episodes is not None else None,
                "rating": rating,
                "description": item.get("description"),
                "image_url": item.get("image_url"),
                "link": item.get("link"),
                "genres": item.get("genres") if isinstance(item.get("genres"), list) else [],
                "tags": item.get("tags") if isinstance(item.get("tags"), list) else [],
                "main_actors": item.get("main_actors") if isinstance(item.get("main_actors"), list) else [],
            }

            if embeddings is not None:
                # PostgREST/pgvector expects vector as string "[x,y,z,...]"
                row["embedding"] = "[" + ",".join(str(float(x)) for x in embeddings[idx]) + "]"
            if umap_2d is not None:
                row["umap_x"] = float(umap_2d[idx, 0])
                row["umap_y"] = float(umap_2d[idx, 1])

            rows.append(row)

        r = requests.post(f"{api}/dramas", json=rows, headers=headers, timeout=60)
        if not r.ok:
            print(f"Batch error at {start}: {r.status_code} {r.text}", file=sys.stderr)
            sys.exit(1)
        inserted += len(rows)
        print(f"Inserted {inserted} / {len(items)}")

    print("Done. Total inserted:", inserted)


if __name__ == "__main__":
    main()
